sw/host/ot_certs/src/asn1/dice_tcb.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use num_bigint_dig::BigUint;
+
+use crate::asn1::builder::Builder;
+use crate::asn1::x509::X509;
+use crate::asn1::{Oid, Tag};
+use crate::template::{FirmwareId, Flags, Value};
+
+#[derive(Debug, Clone)]
+pub struct DiceTcbInfo<'a> {
+    pub vendor: &'a Option<Value<String>>,
+    pub model: &'a Option<Value<String>>,
+    pub version: &'a Option<Value<String>>,
+    pub svn: &'a Option<Value<BigUint>>,
+    pub layer: &'a Option<Value<BigUint>>,
+    pub index: &'a Option<Value<BigUint>>,
+    pub fwids: &'a Option<Vec<FirmwareId>>,
+    pub flags: &'a Option<Flags>,
+    pub vendor_info: &'a Option<Value<Vec<u8>>>,
+    pub tcb_type: &'a Option<Value<Vec<u8>>>,
+}
+
+impl DiceTcbInfo<'_> {
+    // From the DICE specification:
+    // https://trustedcomputinggroup.org/wp-content/uploads/DICE-Attestation-Architecture-r23-final.pdf
+    //
+    // tcg OBJECT IDENTIFIER ::= {2 23 133}
+    // tcg-dice OBJECT IDENTIFIER ::= { tcg platformClass(5) 4 }
+    // tcg-dice-TcbInfo OBJECT IDENTIFIER ::= {tcg-dice 1}
+    // DiceTcbInfo ::== SEQUENCE {
+    //     vendor [0] IMPLICIT UTF8String OPTIONAL,
+    //     model [1] IMPLICIT UTF8String OPTIONAL,
+    //     version [2] IMPLICIT UTF8String OPTIONAL,
+    //     svn [3] IMPLICIT INTEGER OPTIONAL,
+    //     layer [4] IMPLICIT INTEGER OPTIONAL,
+    //     index [5] IMPLICIT INTEGER OPTIONAL,
+    //     fwids [6] IMPLICIT FWIDLIST OPTIONAL,
+    //     flags [7] IMPLICIT OperationalFlags OPTIONAL,
+    //     vendorInfo [8] IMPLICIT OCTET STRING OPTIONAL,
+    //     type [9] IMPLICIT OCTET STRING OPTIONAL
+    // }
+    // FWIDLIST ::== SEQUENCE SIZE (1..MAX) OF FWID
+    //     FWID ::== SEQUENCE {
+    //     hashAlg OBJECT IDENTIFIER,
+    //     digest OCTET STRING
+    // }
+    // OperationalFlags ::= BIT STRING {
+    //     notConfigured (0),
+    //     notSecure (1),
+    //     recovery (2),
+    //     debug (3)
+    // }
+    pub fn push_extension<B: Builder>(&self, builder: &mut B) -> Result<()> {
+        X509::push_extension(builder, &Oid::DiceTcbInfo, false, |builder| {
+            builder.push_seq(Some("dice_tcb_info".into()), |builder| {
+                if let Some(vendor) = self.vendor {
+                    builder.push_string(
+                        Some("dice_vendor".into()),
```

Comment:
We are having way too many allocation for these strings, and just have them discarded as name hints. Any reason it's `Option<String>` instead of `Option<&str>`? (If they are hints then probably even `&str` would be okay?)

Reply:
No particular reason, I think it should be fine to change it to `Option<&str>`. We could change it to just `&str` but that would require a convention for no hint (I guess `""`) so not sure it's much clearer than an `Option`.